R/look_up.R

```diff
@@ -0,0 +1,371 @@
+      }
+    }
+
+    if ("min_ma" %in% colnames(int_key)) {
+      if (!is.numeric(int_key$min_ma)) {
+        stop('`int_key$min_ma` needs to be of type `numeric`')
+      }
+    }
+  }
+
+  #=== Preparation ===
+
+  # save early and late int columns for easier handling
+  early <- occdf[, early_interval]
+  late <- occdf[, late_interval]
+
+  # if there are missing values in `late`, fill them in from `early`
+  replace_pattern <- c("", " ")
+  replace_ind <- which(vapply(late, function(x) {
+    x %in% replace_pattern | is.na(x)
+  }, logical(1)))
+  late[replace_ind] <- early[replace_ind]
+  # in this case, display a warning
+  if(length(replace_ind) >= 1) {
+    warning('`NA`, `""` or `" "` entries from `late_interval` have been
+            filled in with the corresponding `early_interval` entries')
+  }
+
+  # add columns to output data frame
+  occdf$early_stage <- rep(NA_character_, nrow(occdf))
+  occdf$late_stage <- rep(NA_character_, nrow(occdf))
+  if ("max_ma" %in% colnames(int_key) || is.character(assign_with_GTS)) {
+    occdf$interval_max_ma <- rep(NA_real_, nrow(occdf))
+    }
+  if (("max_ma" %in% colnames(int_key) &
+      "min_ma" %in% colnames(int_key)) || is.character(assign_with_GTS)) {
+    occdf$interval_mid_ma <- rep(NA_real_, nrow(occdf))
+    }
+  if ("min_ma" %in% colnames(int_key) || is.character(assign_with_GTS)) {
+    occdf$interval_min_ma <- rep(NA_real_, nrow(occdf))
+    }
+
+  ## early stages unique entries
+  early_unique <- unique(early)
+
+  ## late stages unique entries
+  late_unique <- unique(late)
+
+  #=== Assignment of stages based on look-up table ===
+  if(is.data.frame(int_key)) {
+    # early stage
+    # find assignable intervals
+    assign_ind1 <- vapply(early_unique, function(x) {
+      x %in% int_key$interval_name
+    }, FUN.VALUE = logical(1L))
+    assign1 <- early_unique[assign_ind1]
+    # loop through assignable intervals and assign early stages
+    for (i in seq_len(length(assign1))) {
+      # early stage
+      occdf$early_stage[early == assign1[i]] <-
+        int_key$early_stage[int_key$interval_name == assign1[i]]
+      # max_ma
+      if ("max_ma" %in% colnames(int_key)) {
+        occdf$interval_max_ma[early == assign1[i]] <-
+          int_key$max_ma[int_key$interval_name == assign1[i]]
+      }
+    }
+
+    # late stage
+    # find assignable intervals
+    assign_ind2 <- vapply(late_unique, function(x) {
+      x %in% int_key$interval_name
+    }, FUN.VALUE = logical(1L))
+    assign2 <- late_unique[assign_ind2]
+    # loop through assignable intervals and assign late stages
+    for (i in seq_len(length(assign2))) {
+      #late stage
+      occdf$late_stage[late == assign2[i]] <-
+        int_key$late_stage[int_key$interval_name == assign2[i]]
+      # min_ma
+      if ("min_ma" %in% colnames(int_key)) {
+        occdf$interval_min_ma[late == assign2[i]] <-
+          int_key$min_ma[int_key$interval_name == assign2[i]]
+      }
+    }
+
```

Comment:
Tested and worked on my own dataset. Nice one!

Reply:
That is great to hear :-)